### PR TITLE
Issue: 17964: Run exclude before and after dependencySubstitution to exclude rewritten dependencies

### LINKED ISSUE #17964: The exclude won't be able to exclude rewritten dependency
Even if I add an exclude rule to exclude "com.google.guava:guava", if other dependencies are rewritten to "com.google.guava:guava", I will still get it in the dependency graph. 

Take the following project for example:
```
plugins {
    id 'java'
}

repositories {
    mavenCentral()
}

configurations.all {
    resolutionStrategy.dependencySubstitution {
        substitute module("com.jcraft:jsch") using module("com.google.guava:guava:20.0") because "No reason, just for test"
    }
    exclude group: "com.google.guava", module: "guava"
}

dependencies {
    implementation group: 'org.apache.hadoop', name: 'hadoop-common', version: '2.6.0'
}
```

I already added an exclude rule: `exclude group: "com.google.guava", module: "guava"`

However, I still get guava in my dependency graph
### Expected Behavior

If I exclude a dependency, I should not have it in my dependency graph anymore. No matter what. 

### Current Behavior

clone this repo: https://github.com/attix-zhang/test-gradle-exclude-with-substitution

and then run `./gradlew dependencyInsight --dependency com.google.guava:guava --configuration compileClasspath`,

You will still get guava in the compileClasspath. 
```
➜  test-gradle-exclude-with-substitution git:(main) ./gradlew dependencyInsight --dependency com.google.guava:guava --configuration compileClasspath

> Task :dependencyInsight
com.google.guava:guava:20.0 (selected by rule)
   variant "compile" [
      org.gradle.status              = release (not requested)
      org.gradle.usage               = java-api
      org.gradle.libraryelements     = jar (compatible with: classes)
      org.gradle.category            = library

      Requested attributes not found in the selected variant:
         org.gradle.dependency.bundling = external
         org.gradle.jvm.environment     = standard-jvm
         org.gradle.jvm.version         = 8
   ]

com.jcraft:jsch:0.1.42 -> com.google.guava:guava:20.0
\--- org.apache.hadoop:hadoop-common:2.6.0
     \--- compileClasspath

A web-based, searchable dependency report is available by adding the --scan option.

BUILD SUCCESSFUL in 1s
```
### Context

Our company has some dependency substitution rules and every project is required to use. With this bug, each individual project is unable to exclude dependencies cleanly. 

### Steps to Reproduce 

```
git clone https://github.com/attix-zhang/test-gradle-exclude-with-substitution.git
cd test-gradle-exclude-with-substitution
./gradlew dependencyInsight --dependency com.google.guava:guava --configuration compileClasspath
```

### Your Environment

* Machine: Macbook Big Sur 11.5.1
* Java: 
```
openjdk version "1.8.0_292"
OpenJDK Runtime Environment (Zulu 8.54.0.21-CA-macosx) (build 1.8.0_292-b10)
OpenJDK 64-Bit Server VM (Zulu 8.54.0.21-CA-macosx) (build 25.292-b10, mixed mode)
```
 * Build scan URL:  https://scans.gradle.com/s/23jewesslbjwq

### PR DESCRIPTION
**Pull request by attix-zhang**
Apply the exclude spec before and after dependency substitution to
ensure that redirected dependency will be excluded correctly.

Signed-off-by: Attix Zhang <attix.zhang.cs@gmail.com>

Fixes #17964 

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by ljacomet** [comment]
@bot-gradle test QF

**Comment by ljacomet** [comment]
@attix-zhang Thanks for the contribution!

I do not know why you did not see that checkbox. But it should not be a problem in this case.

I agree that we can cache the substitution result, so all good there.

I have started builds to verify the change has no unexpected consequence.

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by attix-zhang** [comment]
Looks to me, all tests passed! (8 tests passed with re-try, not sure why)

**Comment by ljacomet** [comment]
@bot-gradle test and merge

**Comment by ljacomet** [comment]
Indeed, all tests passed, sending it for merge.

Thanks again for the contribution!

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by attix-zhang** [comment]
Is there any chance that we can backport it to the previous version such as Gradle 6.x? Or the recommended way is to upgrade the Gradle to 7.3+?

**Comment by attix-zhang** [comment]
Oh yeah. Thank you! @ljacomet

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java [hunk 1]
```diff
@@ -434,7 +434,6 @@ private void visitDependencies(ExcludeSpec resolutionFilter, Collection<EdgeStat
         try {
             collectAncestorsStrictVersions(incomingEdges);
             for (DependencyState dependencyState : dependencies(resolutionFilter)) {
-                dependencyState = maybeSubstitute(dependencyState, resolveState.getDependencySubstitutionApplicator());
```
[the remark this answers is not in the document]
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java:437]: `dependencies(resolutionFilter)` now performs the substitution. So the below method still apply to the substituted dependency.

[the remark this answers is not in the document]
Comment (mmorrisontx) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java:437]: Doh, for some reason I had it in my mind that `cacheFilteredDependencyStates` was only called later on. Thanks for clarification 😄

**Pull request merged**